### DISCUSSION TITLE
db65 debugger

### DISCUSSION THREAD

**Discussion by pm100** [cc65/cc65#2323] [General]
Not really an issue but a pointer to a tool I make over the last few weeks. I got fed up having to use VICE as my debugger for working on code. So I built a debugger around the sim65 core engine.

It understands sinm65 behavior, all the PV hooks, VICE symbols etc. It is much easier to use than firing up VICE

Its here https://github.com/pm100/db65

There is one release that has binaries for mac, linux and windows

https://github.com/pm100/db65/releases

I would welcome any and all input.

If people want to add features/ fix bugs... fire away. Heads up tho, its written in rust. 

I had to make 2 tiny changes to sim65/6502.c which I would like to PR back if possible

**Comment by pm100** [comment]
even if no further interest here I would appreciate it if you could add a reference to it in the readme or docs alongside the mentions of using VICE as a debugger

**Comment by pm100** [comment]
I intend to make distributable packages available for db65 (homebrew, chocolatey..) . It compiles 6502.c into the program. I have updated my LICENSE file to include a copy of the copyright from 6502.c to acknowledge this

I also added " Ullrich von Bassewitz, Christian Krueger and others in cc65 project" as authors in the various packe templates. These are the two names in the 6502.c file

**Comment by pm100** [comment]
@mrdudz - I am at a dead end here. I would like to put this in the docs along side the pointers to other debuggers but dont want to do that unless at least somebody else has tried it out. But nobody will try it out if they dont know it exists

Given that you are working on float support you are exactly the kind of dev I was targetting - could you try it out. There are realeases on the gh homepage., together with a short demo of it in use, the integrated help is ok ish.

**Comment by mrdudz** [comment]
I am the wrong person to ask, sorry - if you look at that float branch, you'll notice its sprinkled with log calls all over it. Thats how i debug :)

**Comment by rofl0r** [comment]
if you'd write it in C instead of rust, it could be added to the cc65 suite eventually. also would make it easier for folks to try it out, instead of running 3rd party binaries on their machine.

apart from that, to be really useful a generic debugger for 6502 naturally should support all machines with a 6502 in them (at least the subset supported by cc65) so it would need to become something like a multi-machine emulator. a huge task, so in the end it's probably smarter to invest the effort into improving the debuggers in existing emulators.

**Comment by pm100** [comment]
It's designed specifically for people working in cc65 internals. I was doing a complete rewrite of the printf core so that it's 100% compliant.

Btw I did an article on how it's built https://medium.com/rustaceans/interfacing-c-with-rust-to-build-a-debugger-93bb1c8da6f6

________________________________

**Comment by kugelfuhr** [comment]
> apart from that, to be really useful a generic debugger for 6502 naturally should support all machines with a 6502 in them (at least the subset supported by cc65) so it would need to become something like a multi-machine emulator. a huge task, so in the end it's probably smarter to invest the effort into improving the debuggers in existing emulators.

While I agree with your comments regarding the usage of rust, a generic debugger can be quite helpful when writing / debugging routines that do not depend on the platform. A prominent example are runtime and library routines for the compiler itself.

To be really useful, I would expect such a debugger to have at least some features not available in the VICE monitor.

An incomplete wish list might be:

- full support for all symbols from the assembler file for easy navigation in the code
- source code step through or at least source code line information on errors
- automatic memory protection based on segment information to detect memory overwrites / invalid jumps
- switchable "exceptions" in case of irregular behavior like 
  * illegal opcodes     
  * execution of code in the zeropage/stack/invalid segments
  * reading uninitialized memory
  * writing to non existing memory
  * hardware stack over-/underflow
  * software stack over-/underflow
  * memory access via NULL (check zp argument for (zp),y)
  * ...
- watches for named variables (local ones is probably difficult)
- breakpoints on writes and reads for named variables
- ...

**Comment by pm100** [comment]
It's got a lot of those features already, at the assembler level. Implementing them for c level would require huge changes to all the cc65 tool chain. I am certainly up for that if I know that it would be used. I only work at assembler level

________________________________

**Comment by pm100** [comment]
I am looking for another emulator that I can plug it into. Not vice, that already has a debugger. Any other that people could suggest.ideally written in c since that's relatively easy to hook into

I have a branch where I made it run kim1 binaries, but using the sim65 instruction engine.

________________________________

**Comment by polluks** [comment]
How about Frodo <http://frodo.cebix.net>?

Am Fr., 5. Jan. 2024 um 22:36 Uhr schrieb pm100 ***@***.***>:

**Comment by pm100** [comment]
Thanks for the pointer. A product almost as old as cc65 itself I see. I will take a look

________________________________

**Comment by groessler** [comment]
Or Atari800

**Comment by kugelfuhr** [comment]
> It's got a lot of those features already, at the assembler level.

Now that's impressive. Especially since the necessary information is neither available from the binary file header nor from the VICE label file.

Being curious I wanted to have a look at it, but none of the binaries you've provided is runnable here. I would be required to download, compile & install the rust compiler itself, before being able to compile your stuff. And this is not going to happen. The preferred installation procedure (rustup) works around any package management. Which is something I hate because it usually means that it cannot be uninstalled without leaving traces everywhere.

So I'm out of this, sorry.

**Comment by pm100** [comment]
Could you say why the binaries aren't runnable. I tested them all out.

**Reply by kugelfuhr** [reply]
> Could you say why the binaries aren't runnable. I tested them all out.

```
./db65: /lib/x86_64-linux-gnu/libm.so.6: version `GLIBC_2.35' not found (required by ./db65)
./db65: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.33' not found (required by ./db65)
./db65: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.32' not found (required by ./db65)
./db65: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.34' not found (required by ./db65)
```

**Reply by groessler** [reply]
You need a newer glibc version.

**Reply by groessler** [reply]
or compile db65 using an older version

**Reply by mrdudz** [reply]
You'll end up with rustup upgrading your libc :)

**Reply by pm100** [reply]
@kugelfuhr TY - I will fix this

**Comment by mrdudz** [comment]
BTW, you could also hook your debugger up to VICE (using the binary remote protocol)

**Comment by pm100** [comment]
@kugelfuhr @groessler @mrdudz There is now a MUSL binary in the releases section , this has no external dependencies

TIL why rust product builders use MUSL , I had seen it but did not know why

**Comment by kugelfuhr** [comment]
Ok, so here is a short review:

1. The MUSL binary works.
2. The commands are somewhat confusing. Why does "load_code" load the binary, but to load the symbols you need just "symbols"?
3. The help text should be sorted somehow. Currently it is jumbled so one must read through the whole text when searching things.
4. The syntax is as confusing as the VICE syntax which is bad. Why does it need a dot before the identifiers?
5. Error messages are confusing. I couldn't make anything out of
```
>> break _main
invalid digit found in string
>> break =_main
Variable identifier is not bound to anything by context: "_main".
```
6. For the disassembly it would sometimes be helpful to see the actual code bytes.
7. The symbols are actually not very helpful. The following is the disassembly of "pushax".  Do you really believe it is decrementing `__EXEHDR__`?
```
.pushax:
028c:       pha   
028d:       lda   .sp
028f:       sec   
0290:       sbc   #$02 
0292:       sta   .sp
0294:       bcs   .@L1
0296:       dec   .__EXEHDR__
.@L1:
```
8. With the same identifiers in more than one module, the disassembler just picks one, so "print foo" will always display the same "foo" even if there are several in the program. This is especially bad for labels. Using "break .L2" will choose any label named "L2" by random and ignores all others.
9. "finish" does not seem to run to the end of the function if it ends in a `jmp` instead of an `rts`.
10. I wasn't able to have the debugger catch any invalid operations. Writing so some random memory outside of any segment and with checks enabled does not seem to have an effect. Jumping to some random locations from my code sends the debugger in an endless loop.
11. I haven't found a way to stop a runaway program. Pressing "Ctrl-C" while the program is in a loop simply terminates the debugger.
12. From my list above, only two features are partially(!) implemented. Namely watches and breakpoints for named variables. The use is severely limited by the problems with label handling. This is far from your claim "It's got a lot of those features already".

Summary: The debugger mimics the VICE monitor but does not really improve on it. By using the VICE label file, it inherits the same problems with regard to labels. It does even implement the nasty "dot before identifier" syntax. Compared to VICE it has some disadvantages like not catching any invalid behavior.

**Comment by pm100** [comment]
@kugelfuhr Thanks for spending time looking at it. 


ctrl-c is on the todo list (in my head)

the memory check checks for reads of unwritten data, not access outside segments, I will see if reading the ld65 map out can help here

No I do not believe it is decrementing EXEHDR,  :-) , its just that EXHDR happens to be located at sp+1 and this is the first symbol I found matching that address. 

I should read the -vm memory map , this will improve the symbol handling. and enable better boundary checkinig

I found it very useful to have . at the start of a symbol in commands because its syntactically clear, but it turns out i probably dont need it - no symbol starts with a digit, no numbers start with a character . Will remove

Again ty. Having a new pair of eyes on something is always useful. 

Note that it was built specifically to avoid having to recompile and fire up vice in order to debug code in libsrc/runtime and libsrc/common, ie non platform specific stuff (I needed it during rewrite of _printf.s). It can be used purely from the command line. It is sim65 core engine but with a debugger wrapped around it.

**Reply by pm100** [reply]
@kugelfuhr there is a new version based on your feedback. I think it addresses all your issues see https://github.com/cc65/cc65/discussions/2394
